src/shared/particle_dynamics/slender_structure_math.cpp

```diff
@@ -0,0 +1,264 @@
+								  - sin_rotation_0 * (dpseudo_n_d2t[1] + cos_rotation_0 * angular_vel[0] * angular_vel[0]);
+
+            return Vec2d(angle_vel_dt_0, 0.0);
+		}
+		//=================================================================================================//
+		Vec3d getRotationFromPseudoNormalForFiniteDeformation(const Vec3d &dpseudo_n_d2t, const Vec3d &rotation, const Vec3d &angular_vel, Real dt)
+		{
+			Real sin_rotation_0 = sin(rotation[0]);
+			Real cos_rotation_0 = cos(rotation[0]);
+			Real sin_rotation_1 = sin(rotation[1]);
+			Real cos_rotation_1 = cos(rotation[1]);
+
+			Real rotation_0_a = -(dpseudo_n_d2t[2] * cos_rotation_1 + dpseudo_n_d2t[0] * sin_rotation_1
+								  + angular_vel[1] * angular_vel[1] * cos_rotation_0
+								  + angular_vel[0] * angular_vel[0] * cos_rotation_0);
+			Real rotation_0_b = sin_rotation_0 * angular_vel[0] * angular_vel[0] - dpseudo_n_d2t[1];
+			Real angle_vel_dt_0 = sin_rotation_0 * rotation_0_a + cos_rotation_0 * rotation_0_b;
+
+			Real rotation_1_a = dpseudo_n_d2t[0] * cos_rotation_1 - dpseudo_n_d2t[2] * sin_rotation_1
+								+ 2.0 * angular_vel[1] * angular_vel[0] * sin_rotation_0;
+			Real rotation_1_b1 = dpseudo_n_d2t[0] * cos_rotation_0
+								 + angular_vel[1] * angular_vel[1] * cos_rotation_0 * cos_rotation_0 * sin_rotation_1
+								 + angular_vel[0] * angular_vel[0] * sin_rotation_1
+								 - dpseudo_n_d2t[1] * sin_rotation_1 * sin_rotation_0
+								 + 2.0 * angular_vel[1] * angular_vel[0] * cos_rotation_1 * cos_rotation_0 * sin_rotation_0;
+			Real rotation_1_b2 = -(dpseudo_n_d2t[2] * cos_rotation_0
+								   + angular_vel[1] * angular_vel[1] * cos_rotation_1 * cos_rotation_0 * cos_rotation_0
+								   + angular_vel[0] * angular_vel[0] * cos_rotation_1
+								   - dpseudo_n_d2t[1] * cos_rotation_1 * sin_rotation_0
+								   - 2.0 * angular_vel[1] * angular_vel[0] * cos_rotation_0 * sin_rotation_1 * sin_rotation_0);
+			Real angle_vel_dt_1 = rotation_1_a * rotation_1_a * (rotation_1_b1 * cos_rotation_1 + rotation_1_b2 * sin_rotation_1)
+								  / (rotation_1_b1 * rotation_1_b1 + rotation_1_b2 * rotation_1_b2 + Eps);
+					
+			return Vec3d(angle_vel_dt_0, angle_vel_dt_1, 0.0);
+		}
+		//=================================================================================================//
+		Vec2d getRotationFromPseudoNormalForSmallDeformation(const Vec2d &dpseudo_n_d2t, const Vec2d &rotation, const Vec2d &angular_vel, Real dt)
+		{
+			return Vec2d(dpseudo_n_d2t[0], 0);
+		}
+		//=================================================================================================//
+        Vec3d getRotationFromPseudoNormalForSmallDeformation(const Vec3d &dpseudo_b_n_d2t, const Vec3d &dpseudo_n_d2t, const Vec3d &rotation, const Vec3d &angular_vel, Real dt)
+		{
+            return Vec3d(0.0, dpseudo_n_d2t[0], 0.0);
+		}
+		//=================================================================================================//
```

Comment:
The slender structure is only built up in 3D formulation, right? If it is, please clean all 2D code.
And why is only one angle used to update pseudo-normal?